addons/hr_payroll/models/hr_payslip.py

```diff
@@ -0,0 +1,581 @@
+        """
+        @param employee: browse record of employee
+        @param date_from: date field
+        @param date_to: date field
+        @return: returns the ids of all the contracts for the given employee that need to be considered for the given dates
+        """
+        #a contract is valid if it ends between the given dates
+        clause_1 = ['&',('date_end', '<=', date_to),('date_end','>=', date_from)]
+        #OR if it starts between the given dates
+        clause_2 = ['&',('date_start', '<=', date_to),('date_start','>=', date_from)]
+        #OR if it starts before the date_from and finish after the date_end (or never finish)
+        clause_3 = ['&',('date_start','<=', date_from),'|',('date_end', '=', False),('date_end','>=', date_to)]
+        clause_final =  [('employee_id', '=', employee.id),'|','|'] + clause_1 + clause_2 + clause_3
+        return self.env['hr.contract'].search(clause_final).ids
+
+    @api.multi
+    def compute_sheet(self):
+        HrPayslipLine = self.env['hr.payslip.line']
+        IrSequence = self.env['ir.sequence']
+        for payslip in self:
+            number = payslip.number or IrSequence.next_by_code('salary.slip')
+            #delete old payslip lines
+            old_slipline_ids = HrPayslipLine.search([('slip_id', '=', payslip.id)])
+            #old_slipline_ids
+            if old_slipline_ids:
+                old_slipline_ids.unlink()
+            if payslip.contract_id:
+                #set the list of contract for which the rules have to be applied
+                contract_ids = [payslip.contract_id.id]
+            else:
+                #if we don't give the contract, then the rules to apply should be for all current contracts of the employee
+                contract_ids = self.get_contract(payslip.employee_id, payslip.date_from, payslip.date_to)
+            lines = [(0,0,line) for line in payslip.get_payslip_lines(contract_ids)]
+            payslip.write({'line_ids': lines, 'number': number})
+        return True
+
+    @api.multi
+    def get_worked_day_lines(self, contract_ids, date_from, date_to):
+        """
+        @param contract_ids: list of contract id
+        @return: returns a list of dict containing the input that should be applied for the given contract between date_from and date_to
+        """
+        def was_on_leave(employee_id, datetime_day):
+            res = False
+            day = datetime_day.strftime("%Y-%m-%d")
+            holiday_id = self.env['hr.holidays'].search([('state','=','validate'),('employee_id','=',employee_id),('type','=','remove'),('date_from','<=',day),('date_to','>=',day)], limit=1)
+            if holiday_id:
+                res = holiday_id.holiday_status_id.name
+            return res
+
+        res = []
+        HrContract = self.env['hr.contract']
+        ResourceCalender = self.env['resource.calendar']
+        for contract in HrContract.browse(contract_ids):
```

Comment:
cannot we use filtered here....
HrContract.browse(contract_ids).filtered('working_hours')